### PR TITLE
[SYCLomatic] Fix the incorrect extra arguments inserting location

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by ShengchenJ — APPROVED** [review]

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -4344,8 +4346,39 @@ DeviceFunctionDecl::DeviceFunctionDecl(
   buildTextureObjectParamsInfo(FTL.getParams());
 }
 std::shared_ptr<DeviceFunctionInfo>
-DeviceFunctionDecl::LinkUnresolved(const UnresolvedLookupExpr *ULE) {
-  return LinkDeclRange(ULE->decls(), getFunctionName(ULE));
+DeviceFunctionDecl::LinkUnresolved(const UnresolvedLookupExpr *ULE,
+                                   std::optional<unsigned int> NumArgs) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:4349]:
```suggestion
                                  unsigned int NumArgs) {
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -4133,7 +4134,8 @@ void CallFunctionExpr::setFuncInfo(std::shared_ptr<DeviceFunctionInfo> Info) {
   }
   FuncInfo = Info;
 }
-void CallFunctionExpr::buildCalleeInfo(const Expr *Callee) {
+void CallFunctionExpr::buildCalleeInfo(const Expr *Callee,
+                                       std::optional<unsigned int> NumArgs) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:4137]:
```suggestion
                                       unsigned int NumArgs) {
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -3925,10 +3925,11 @@ void CallFunctionExpr::buildCallExprInfo(const CXXConstructExpr *Ctor) {
       0, SM, DpctGlobalInfo::getContext().getLangOpts()));
 }
 
-void CallFunctionExpr::buildCallExprInfo(const CallExpr *CE) {
+void CallFunctionExpr::buildCallExprInfo(const CallExpr *CE,
+                                         std::optional<unsigned int> NumArgs) {
   if (!CE)
     return;
-  buildCalleeInfo(CE->getCallee()->IgnoreParenImpCasts());
+  buildCalleeInfo(CE->getCallee()->IgnoreParenImpCasts(), NumArgs);
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
  buildCalleeInfo(CE->getCallee()->IgnoreParenImpCasts(), CE->getNumArgs());
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 4]
```diff
@@ -4344,8 +4346,39 @@ DeviceFunctionDecl::DeviceFunctionDecl(
...
 std::shared_ptr<DeviceFunctionInfo>
-DeviceFunctionDecl::LinkUnresolved(const UnresolvedLookupExpr *ULE) {
-  return LinkDeclRange(ULE->decls(), getFunctionName(ULE));
+DeviceFunctionDecl::LinkUnresolved(const UnresolvedLookupExpr *ULE,
+                                   std::optional<unsigned int> NumArgs) {
+  std::shared_ptr<DeviceFunctionInfo> Info;
+  DeclList List;
+
+  for (auto *D : ULE->decls()) {
+    if (NumArgs) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:4352]:
```suggestion

```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 5]
```diff
@@ -4344,8 +4346,39 @@ DeviceFunctionDecl::DeviceFunctionDecl(
...
+      if (const FunctionDecl *FD = dyn_cast<FunctionDecl>(D)) {
+        if (NumArgs.value() >= FD->getMinRequiredArguments() &&
+            NumArgs.value() <= FD->getNumParams()) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: use a vector to save the filtered decls, pass it to LinkDeclRange

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 6]
```diff
@@ -4344,8 +4346,39 @@ DeviceFunctionDecl::DeviceFunctionDecl(
...
+          LinkDecl(D, List, Info);
+        }
+      }
+      if (const FunctionTemplateDecl *FTD = dyn_cast<FunctionTemplateDecl>(D)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:4362]:
```suggestion
      else if (const FunctionTemplateDecl *FTD = dyn_cast<FunctionTemplateDecl>(D)) {
```

Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Merge two if stmts

**Review by ziranzha — APPROVED** [review]

**Pull request merged**